### PR TITLE
[css-scroll-snap] choose closest snap target if neither are visible

### PR DESCRIPTION
**Pull request by nmoucht**
#### 0efae6b49e5cb1c1287480ae860dbde5a3d269b1
<pre>
[css-scroll-snap] choose closest snap target if neither are visible
<a href="https://bugs.webkit.org/show_bug.cgi?id=245164">https://bugs.webkit.org/show_bug.cgi?id=245164</a>
&lt;rdar://99897147&gt;

Reviewed by Wenson Hsieh.

When taking the closest snap target for each axes results in neither target being visible, pick
the closest target to the user scroll position.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-to-visible-areas-both-expected.txt:
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::offsetHasVisibleSnapArea):
(WebCore::findCompatibleSnapArea):
(WebCore::adjustPreviousAndNextForOnScreenSnapAreas):
(WebCore::ensureVisibleTarget):
(WebCore::LayoutScrollSnapOffsetsInfo::closestSnapOffset const):
(WebCore::FloatScrollSnapOffsetsInfo::closestSnapOffset const):
(WebCore::hasCompatibleSnapArea): Deleted.

Canonical link: <a href="https://commits.webkit.org/254982@main">https://commits.webkit.org/254982@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/8400167d9a974fe7c81f44be0eb97a7e21ce2cba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90917 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35488 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/100219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94923 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33989 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/83260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96573 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/27089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/83260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/83260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35069 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/3471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36647 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38574 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nt1m) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: nit: space before =

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nt1m) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: nit: superfluous blank line

**Review by nt1m — COMMENTED** [review]
I'll leave it to Martin for the actual review.

**Comment by nt1m** [comment]
(also please fix the typo in the commit message: closet -> closest)

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nt1m) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: nit: remove this logging

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (whsieh) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Nit - perhaps `indexOfCompatibleSnapArea` or `findCompatibleSnapArea`?

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (whsieh) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Nit - might be cleaner as

```
for (auto index : snapOffset.snapAreaIndices) {
    …
}
```

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (whsieh) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Nit - space before `*` in `scrollDestinationOffset.x())*`

**Review by whsieh — APPROVED** [review]

**Comment by nt1m** [comment]
Please remove the flaky test annotation you added over the weekend too.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nt1m) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: typo:
```suggestion
static bool offsetHasVisibleSnapArea(const InfoType& info, const SnapOffset<UnitType>& snapOffset, const SnapOffset<UnitType>& snapOffsetOther, ScrollEventAxis axis, const SizeType& viewportSize)
```

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h
Comment (nt1m) [inline Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h]: Should this be removed?

**Review by nt1m — COMMENTED** [review]
Can you undo https://bugs.webkit.org/show_bug.cgi?id=245717 ? 

Also, can you update the changelog file/function list as appropriate?

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h]: Fixed.

**Comment by nmoucht** [comment]
> Please remove the flaky test annotation you added over the weekend too.

Looks like this didn't fix the flaky test (only flaky on mac-AS-debug-wk2). Currently investigating.

**Comment by nt1m** [comment]
> > Please remove the flaky test annotation you added over the weekend too.
> 
> Looks like this didn't fix the flaky test (only flaky on mac-AS-debug-wk2). Currently investigating.

can we make the annotation debug only?

**Comment by nmoucht** [comment]
> > > Please remove the flaky test annotation you added over the weekend too.
> > 
> > 
> > Looks like this didn't fix the flaky test (only flaky on mac-AS-debug-wk2). Currently investigating.
> 
> can we make the annotation debug only?

Actually looks like it failed in Mac-wk2 as well. It passed on wk1 bots but this could be do to flakiness. I'll check the bots once this lands to see what it can be enabled on. I have a fix in mind for this once this patch lands.

**Pull request closed without merging**
